File: OOPs/Constructors/default_copy_constructor.py
# Example (Using the copy Module):
import copy

class Car:
    def __init__(self, brand, year):
        self.brand = brand
        self.year = year

    def display_info(self):
        print(f"Car: {self.brand}, Year: {self.year}")

car1 = Car("Toyota", 2020)
car2 = copy.copy(car1)  # Shallow copy
car2.display_info()



# Example (Manual Copy Constructor):
class CarManualCopy:
    def __init__(self, brand, year):
        self.brand = brand
        self.year = year

    def display_info(self):
        print(f"Car: {self.brand}, Year: {self.year}")

    def copy(self):
        return Car(self.brand, self.year)

car1 = CarManualCopy("Toyota", 2020)
car2 = car1.copy()
car2.display_info()

File: OOPs/Constructors/default_copy_constructor.cpp
#include <iostream>
using namespace std;

class Car {
private:
    string brand;
    int year;

public:
    // Constructor
    Car(string b, int y) : brand(b), year(y) {}
    /*      OR
    Car(string b, int y) {
        brand = b, year = y;
    }
    */

   // Default / Shallow Copy | We can skip writing it cause it will be created autometicly
    Car(Car &obj) {
        this->brand = obj.brand;
        this->year = obj.year;
    }

    // Method to display car details
    void displayInfo() {
        cout << "Car: " << brand << ", Year: " << year << endl;
    }
};

int main() {
    Car car1("Toyota", 2020);  // Normal constructor
    Car car2(car1);          // Default copy constructor is invoked

    car2.displayInfo();  // Output: Car: Toyota, Year: 2020
    return 0;
}
